Travis deploy

werebus opened the pull request:
Use Travis to deploy to Github releases.

Anbranin commented on .travis.yml:
```diff
@@ -0,0 +1,9 @@
+deploy:
+  provider: releases
+  api_key:
+    secure: TQuDynLxEXAfBJeC/dXpXhFzRt8a0DiYFQlfRoKnrAF5P7ZZg0MikEuvjXbwrKA2ArzXVBiFGYEnficjH9zv60ZDHddjDguB4wl72uHqR1lXhgMRzU3s4FTFdFv4g5H7iOOhrqYRh9Sxvgdvf260YxWAMKi8ip6QS1StTi0AEI1d56J+/95aDZZW8aWK9NFuQgZvcm1kBggIvewFKPrDIBEQEWSLeYGu2oT+hmv61s+t+RgUNXhjQ00VuWeHkGHMDzl4ljlJVlC1ZtrnqW9fiYs5VUeJZTX3Ob4e+7rd0noQTyGjaOReLrzuUWLr8MF2fZln9yuiZlhgkvLTmiEdCUgn3BcCFVfri3MgN9bBfuhNclSxVK6qluTPk4eHvPALZSZ ...
```
So it's cool if the API key is public? I don't know about these things, it just "feels" odd to me.

werebus replied:
https://docs.travis-ci.com/user/encryption-keys/

Essentially, this is the encrypted value of the API Key - encrypted with a private key that only Travis has.

werebus replied:
The value, once decrypted, contains an API key that belongs to @umts-buildbot-gismap. It grants write access to public repositories which that user has write access to. And, that user only has write access to this repository.

Anbranin commented on src/index.html:
```diff
@@ -7,15 +7,12 @@
     <link rel="stylesheet" href="index.css">
     <script>
-      // Current path minus everything after the last forward slash
-      var locationPath = location.pathname.replace(/\/[^\/]+$/, "");
```
So you don't have to do this because you moved the assets to `src`?

werebus replied:
It was an intermediate variable whose use was removed below.

werebus replied:
And also, I was wrong.

werebus commented:
Ahem...

![2019-06-21_13-01](https://user-images.githubusercontent.com/240275/59939040-aafdbd00-9424-11e9-9b01-a1e65e54e9a3.png)

werebus commented:
So, the primary change here is organizational:

* `app/` contains the Typescript that will be compiled with `tsc`
* `src/` contains static files that are copied over verbatim
* `out/` contains the "built" site

I also removed the requirements to have anything "globally" installed in Node and added some helper scripts:  `yarn run build`, `yarn run compress`, `yarn run serve`

Travis only _actually_ does the release upload on `master`, so that part is still a little un-tested. More tweaking might be required.

LiamBrandt commented on README.md:
````diff
@@ -1,24 +1,34 @@
+$ yarn run build
+```
+
+The built site will be in the `out/` directory. Serve it with:
````
What do you mean by "serve it"?

werebus replied:
Start a webserver? I can be a bit more explicit, I guess. Although, I think the output is (moderately) explanatory:

```
$yarn run serve
yarn run v1.16.0
$ yarn run static out/
$ /home/matt/projects/GISMap/node_modules/.bin/static out/
serving "out/" at http://127.0.0.1:8080
```

werebus commented:
In my mind, It's about being able to keep separate the build artifact and its sources. If you start out with all the static assets in the `out/` directory and then build the app, you're left with a directory in which _some_ things can be edited by the developer, and _some_ things shouldn't be.  This way, it's easy to say, "don't point your editor at `out/`, it gets destroyed every time you compile the app."

werebus commented:
I took `src/` as a convention from a few static site generators I've used. But, I'm not married to it.

werebus commented:
Also, another advantage of keeping `src/` (or, whatever) and `out/` separate is it makes keeping the compiled js from being committed (in the `.gitignore`) a bit simpler.

Other options:
*  Only gitignore .js and .js.map files, but that precludes adding plain-old js files in (for whatever reason)
*  Explicitly un-ignore `index.html`, etc. Although that adds overhead to adding e.g. a favicon or whatever.

werebus commented on src/index.html:
```diff
@@ -7,15 +7,12 @@
           {
             name: "app",
-            location: location.protocol + "//" + location.host + locationPath + "/app"
+            location: location.origin + "/app"
```
Just realized that this is bad. Won't work if this app is deployed to e.g. a subdirectory.

LiamBrandt commented:
That makes more sense now, especially the separation of `src/` and `out/`.

I still think `static/` would be a better name for `src/`. `src/` is something you compile, which here is not the case. But I also wouldn't mind if we keep it `src/` so long as it's documented in the readme what that folder is and how `yarn run build` uses it.

UmassGIS commented:
Just a suggestion to consider: some project use “dist” folder to indicate that the content is ready for distribution and to distinguish from the source/config/dev files.

From: Liam Brandt <notifications@github.com>
Sent: Monday, June 24, 2019 12:59 PM
To: umts/GISMap <GISMap@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [umts/GISMap] Travis deploy (#57)


@LiamBrandt requested changes on this pull request.

If we aren't going to change src folder to static, can we at least document what app and src are used for when running yarn run build in the readme? That way new developers will immediately know why there is a separation of those folders.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/umts/GISMap/pull/57?email_source=notifications&email_token=AEWNIYIL35CPIAKX2BM7FNDP4D4MBA5CNFSM4H2TE6CKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4OK62A#pullrequestreview-253538152>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEWNIYNZTWHV5ZHCCYXCIALP4D4MBANCNFSM4H2TE6CA>.

werebus commented:
I'm sold on `static/`, but I think I'm going to leave it `out/` rather than `dist/` which, I agree, _is_ a pretty common convention. However, The output directory is arguably an intermediate step in deploying — the final step being a zip file — and also used during development.

werebus commented:
Alright, I'm going to try merging and see if it actually builds a release.

werebus commented:
Tada:  https://github.com/umts/GISMap/releases/tag/untagged-c8b6b494b1ddf7aba115
